chain/client/src/client_actor.rs

```diff
@@ -1784,6 +1782,37 @@ impl ClientActor {
             &self.config_updater,
         )
     }
+
+    /// Checks if the node is syncing its State and applies special logic in that case.
+    /// A node usually ignores blocks that are too far ahead, but in case of a node syncing its state it is looking for 2 specific blocks:
+    /// * The first block of the new epoch
+    /// * The last block of the prev epoch
+    /// Returns whether the node is syncing its state.
+    fn maybe_receive_state_sync_blocks(&mut self, block: &Block) -> bool {
+        match self.client.sync_status {
+            SyncStatus::StateSync(StateSyncStatus { sync_hash, .. }) => {
+                if let Ok(header) = self.client.chain.get_block_header(&sync_hash) {
+                    let block: MaybeValidated<Block> = (*block).clone().into();
+                    // Notice that two blocks are saved differently:
+                    // * save_block() for one block.
+                    // * save_orphan() for another block.
+                    if block.hash() == header.prev_hash() {
+                        // The last block of the previous epoch.
+                        if let Err(err) = self.client.chain.save_block(block) {
+                            error!(target: "client", ?err, "Failed to save a block during state sync");
+                        }
+                    } else if block.hash() == &sync_hash {
+                        // The first block of the new epoch.
+                        if let Err(err) = self.client.chain.save_orphan(block, false) {
+                            error!(target: "client", ?err, "Received an invalid block during state sync");
```

Comment:
minor: in both errors, let's mention which block provides the error (e.g. first block of the new epoch)
We'll be able to distinguish because the errors are different, but we can simplify this detective work in the very beginning